src/Common/src/System/IO/PersistedFiles.Unix.cs

```diff
@@ -0,0 +1,88 @@
+{
+    internal static class PersistedFiles
+    {
+        // If we ever need system persisted data, /etc/dotnet/corefx/
+        private const string TopLevelDirectory = "dotnet";
+        // User persisted data, ~/.dotnet/corefx/
+        private const string TopLevelUserDirectory = "." + TopLevelDirectory;
+        private const string SecondLevelDirectory = "corefx";
+
+        // Do not use the user variables in the static initializer in case $HOME didn't resolve.
+        private static string s_userHomeDirectory;
+        private static string s_userProductDirectory;
+
+        /// <summary>
+        /// Get the location of where to persist information for a particular aspect of the framework,
+        /// such as "cryptography".
+        /// </summary>
+        /// <param name="featureName">The directory name for the feature</param>
+        /// <returns>A path within the user's home directory for persisting data for the feature</returns>
+        internal static string GetUserFeatureDirectory(string featureName)
+        {
+            if (s_userProductDirectory == null)
+            {
+                EnsureUserDirectories();
+            }
+
+            return Path.Combine(s_userProductDirectory, featureName);
+        }
+
+        /// <summary>
+        /// Get the location of where to persist information for a particular aspect of a feature of
+        /// the framework, such as "x509stores" within "cryptography".
+        /// </summary>
+        /// <param name="featureName">The directory name for the feature</param>
+        /// <param name="subFeatureName">The directory name for the sub-feature</param>
+        /// <returns>A path within the user's home directory for persisting data for the sub-feature</returns>
+        internal static string GetUserFeatureDirectory(string featureName, string subFeatureName)
+        {
+            if (s_userProductDirectory == null)
+            {
+                EnsureUserDirectories();
+            }
+
+            return Path.Combine(s_userProductDirectory, featureName, subFeatureName);
+        }
+
+        /// <summary>
+        /// Get the location of where to persist information for a particular aspect of the framework,
+        /// with a lot of hierarchy, such as ["cryptography", "x509stores", "my"]
+        /// </summary>
+        /// <param name="featurePathParts">A non-empty set of directories to use for the storage hierarchy</param>
+        /// <returns>A path within the user's home directory for persisting data for the feature</returns>
+        internal static string GetUserFeatureDirectory(params string[] featurePathParts)
+        {
+            Debug.Assert(featurePathParts != null);
+            Debug.Assert(featurePathParts.Length > 0);
+
+            if (s_userProductDirectory == null)
+            {
+                EnsureUserDirectories();
+            }
+
+            return Path.Combine(s_userProductDirectory, Path.Combine(featurePathParts));
+        }
```

Comment:
Just FYI, after this change goes in, I plan to add ```GetUserTempFeatureDirectory``` methods, and then use them in the libraries that generate temporary files, e.g. Pipes, MemoryMappedFiles, etc.